src/test/TxTests.cpp

```diff
@@ -395,26 +395,49 @@ validateTxResults(TransactionFramePtr const& tx, Application& app,
 
 TxSetResultMeta
 closeLedgerOn(Application& app, uint32 ledgerSeq, int day, int month, int year,
-              std::vector<TransactionFrameBasePtr> const& txs, bool skipValid)
+              std::vector<TransactionFrameBasePtr> const& txs, bool strictOrder)
 {
     return closeLedgerOn(app, ledgerSeq, getTestDate(day, month, year), txs,
-                         skipValid);
+                         strictOrder);
 }
 
+class TxSetFrameStrictOrderForTesting : public TxSetFrame
+{
+  public:
+    TxSetFrameStrictOrderForTesting(Hash const& previousLedgerHash)
+        : TxSetFrame(previousLedgerHash){};
+
+    std::vector<TransactionFrameBasePtr>
+    sortForApply() override
+    {
+        return mTransactions;
+    };
+
+    void sortForHash() override{};
+};
+
 TxSetResultMeta
 closeLedgerOn(Application& app, uint32 ledgerSeq, time_t closeTime,
-              std::vector<TransactionFrameBasePtr> const& txs, bool skipValid)
+              std::vector<TransactionFrameBasePtr> const& txs, bool strictOrder)
 {
-    auto txSet = std::make_shared<TxSetFrame>(
-        app.getLedgerManager().getLastClosedLedgerHeader().hash);
+    std::shared_ptr<TxSetFrame> txSet;
+    auto lclHash = app.getLedgerManager().getLastClosedLedgerHeader().hash;
+    if (strictOrder)
```

Comment:
I'm not sure if we want to do this. Transaction order is defined at the protocol level, we really want to know if this is changing.

Reply:
Some tests want to test a specific sequence of transactions. I don't think they're testing that the apply order is the same one than in previous implementations - that's what historical replays are for?

There are a few other ways to do it with different tradeoffs, what do you suggest we do in that case?

Alternative to this approach are:
* not close ledgers in that case - just use some other test wrapper (so not testing apply order, so seems the same)
* control the apply order from the test. This can be done by hard coding some memo in one of the transactions. We'd need to compute the value of that memo somehow.

Reply:
In general we've done this approach "control the apply order from the test. This can be done by hard coding some memo in one of the transactions. We'd need to compute the value of that memo somehow." It's not particularly difficult if the number of transactions is small.

These tests have alerted me in the past to inadvertent changes to transaction order, although I agree historical replay is the ultimate source of truth on this.

Reply:
mmm, maybe what we should do then is have `REQUIRE` statements in `closeLedgerOn` that enforce that the order passed in is the same than the apply order.

As far as pattern to control the apply order, I thought of a third possibility that might be the least invasive: we can change it by changing the hash of the previous ledger, so changing anything about the previous ledger should do it. It can be close time etc - or even modify something in the genesis ledger like `TESTING_UPGRADE_MAX_TX_SET_SIZE`.

Reply:
I'm open to discussion on this, but... the "modify txs to ensure correct order" approach seems backwards to me. If we want to catch ordering bugs, why not have a separate test that focuses on that? The test that I modified expects a specific order. If it fails due to the wrong order, it's not immediately obvious, because we're checking result codes, and nothing explicitly order-related (whereas a designated test could provide a clear error message).
Besides the fact that it's extra time spent guessing and maintaining the right hash (that works for all protocol versions), I also didn't like the fact that it makes the test more confusing, since we have to mess with seemingly unrelated stuff (memo, prev hash, or whatever we pick).

Reply:
I think you are right @marta-lokhova : we should just add a canary test with some non trivial (multiple transactions) transaction set, and have a hardcoded check that verifies that the hash of that tx set is what we expect.
That way we avoid mixing test purpose.

Reply:
This was updated with an extra apply order test.